### DISCUSSION TITLE
How to save encrypted.ctb after errode?

### DISCUSSION THREAD

**Discussion by msspad** [sn4k3/UVtools#414] [Scripts]
When preparing encrypted.ctb for Phrozen Mini 8k:
1. import from sl1
2. erode with 8 iterations all bottom layers
3. Save to encrypted.ctb
4. error message below

- Is there any solution for Phrozen Mini 8k?
- Same issue when applying operations ElephantFootErode() or AddEmptyFirstLayer() with scripts, see below.
- Changing WaitBeforeCure works fine (except for first layer), other operations also work fine.

> System.NotSupportedException: 
> Currently we don't allow file save/encode/recreation on purpose on this encrypted file format.
> We are against not permitted user data encryption, locks, constrains, restrictions, monopolization and non-sense policies that prevents others from create and lead-forward solutions around technology.
> Full supporting this file format would be step-back for 3D technology and the community overall. Still, we allow to read the file for review, recover data and detect problems to make you a bit more covered from mistakes.
> Please make your shift from this products and help the 3D technology go forward and open!
>    at UVtools.Core.FileFormats.FileFormat.Encode(String fileFullPath, OperationProgress progress) in D:\Tiago\Dropbox\Programming\C#\UVtools\UVtools.Core\FileFormats\FileFormat.cs:line 2626
>    at UVtools.Core.FileFormats.FileFormat.SaveAs(String filePath, OperationProgress progress) in D:\Tiago\Dropbox\Programming\C#\UVtools\UVtools.Core\FileFormats\FileFormat.cs:line 3507
>    at UVtools.WPF.MainWindow.<>c__DisplayClass109_0.<SaveFile>b__0() in D:\Tiago\Dropbox\Programming\C#\UVtools\UVtools.WPF\MainWindow.axaml.cs:line 2069

```C#
private void ElephantFootErode() {
    if (0 < NumErodeInput.Value) {
        // log.AppendLine($"NumErodeInput={NumErodeInput.Value}");
        OperationMorph morph = new(SlicerFile) {
            MorphOperation = OperationMorph.MorphOperations.Erode,
            Iterations = NumErodeInput.Value,
        };
        morph.SelectBottomLayers();

        Progress.Token.ThrowIfCancellationRequested(); // Abort operation, user requested cancellation
        morph.ROI = Operation.ROI; // Copy user selected ROI to my operation
        morph.MaskPoints = Operation.MaskPoints; // Copy user selected Masks to my operation
        if (morph.CanValidate()) {
            morph.Execute(Progress);
        }
    }
}


private void AddEmptyFirstLayer() {
    var firstLayer = SlicerFile.FirstLayer;
    if (firstLayer.NonZeroPixelCount > 1) { // First layer is not blank as it seems, lets create one
        firstLayer = firstLayer.Clone();
        using var mat = EmguExtensions.InitMat(SlicerFile.Resolution);
        var pixelPos = firstLayer.BoundingRectangle.Center();
        mat.SetByte(pixelPos.X, pixelPos.Y, 1); // Print a very fade pixel to ignore empty layer detection
        firstLayer.LayerMat = mat;
        firstLayer.ExposureTime = SlicerFile.SupportsGCode ? 0 : 0.1f;
        firstLayer.SetWaitTimeBeforeCureOrLightOffDelay(0);
        SlicerFile.SuppressRebuildPropertiesWork(() => { SlicerFile.LayerManager.Prepend(firstLayer); });
    }
}
```

**Comment by sn4k3** [comment]
The **Save** is locked ATM for encrypted ctb.
To by pass that you need to convert the file, via File - Convert - Encrypted CTB
You can call the Convert method into script if you wish

**Reply by msspad** [reply]
- I wasn't precise enough: Menu File -> Save is locked. However, I always use Ctrl+s what works fine.
- My SL1s are configured with FILEFORMAT_ENCRYPTED.CTB in PrusaSlicer Notes, so they convert automatically with the Settings->Automation->Auto convert (though there should be modal window asking if overwriting was ok if there is already a ctb file)
- My issue is that once converted to encrypted.ctb, e.g. erroded layers cannot be saved anymore (neither with File->Save, File->"Save as" or Ctrl+s)

**Reply by sn4k3** [reply]
> Menu File -> Save is locked. However, I always use Ctrl+s what works fine.

Both are the same method, Ctrl+S will raise same message. However you are allowed to change parameters and quick save (Not full enconde)

> Auto convert (though there should be modal window asking if overwriting was ok if there is already a ctb file)

Can be a improvement for a future patch.

> My issue is that once converted to encrypted.ctb, e.g. erroded layers cannot be saved anymore (neither with File->Save, File->"Save as" or Ctrl+s)

Yes, as i told, you need to re-convert your encrypted ctb to encrypted ctb to overcome that, is the same as a full save, made in other way...

**Reply by msspad** [reply]
thanks, saving with a script works now. Code snippet is below in case somebody else is interested. 

```C#
string ctbEncryptedExtension = "encrypted.ctb";
var newFilePath = Path.Combine(SlicerFile.FileDirectoryPath, SlicerFile.FilenameNoExt +"."+ctbEncryptedExtension);
var ctbEncryptedFileFormat = FileFormat.FindByExtensionOrFilePath(ctbEncryptedExtension);
SlicerFile.Convert(ctbEncryptedFileFormat, newFilePath);
```

**Reply by sn4k3** [reply]
Heres a more clean alternative, fail-safe:

```C#
if (SlicerFile.GetType() == typeof(CTBEncryptedFile))
{
    // If SlicerFile is already an encrypted ctb
    SlicerFile.Convert(typeof(CTBEncryptedFile), SlicerFile.FileFullPath);
}
else
{
    // If you convert from other file thats not ctb encrypted
    // You dont need .encrypted.ctb virtual ext because you strict passing the file type to the converter
    var newFilePath = Path.Combine(SlicerFile.FileDirectoryPath, $"{SlicerFile.FilenameNoExt}.ctb");
    SlicerFile.Convert(typeof(CTBEncryptedFile), newFilePath);
}
```
